stan/math/rev/core/vari.hpp

```diff
@@ -300,12 +314,115 @@ class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
+    adj_.setZero();
+  }
+  /**
+   * Implimentation for views which performs a no-op
+   */
+  template <typename Checker = T, require_t<std::is_const<Checker>>* = nullptr>
+  inline void set_zero_adjoint_impl() {}
+
   /**
    * Set the adjoint value of this variable to 0.  This is used to
    * reset adjoints before propagating derivatives again (for
    * example in a Jacobian calculation).
    */
-  inline void set_zero_adjoint() final { adj_.setZero(); }
+  inline void set_zero_adjoint() final { set_zero_adjoint_impl(); }
+
+  /**
+   * A block view of the underlying Eigen matrices.
+   * @param i Starting row of block.
+   * @param j Starting columns of block.
+   * @param p Number of rows to return.
+   * @param q Number of columns to return.
+   */
+  inline const auto block(Eigen::Index i, Eigen::Index j, Eigen::Index p,
+                          Eigen::Index q) const {
+    const auto& val_block = val_.block(i, j, p, q);
+    const auto& adj_block = adj_.block(i, j, p, q);
+    return vari_value<const PlainObject, Eigen::OuterStride<>>(
+        val_block, adj_block, val_block.outerStride(), adj_block.outerStride());
+  }
+
+  /**
+   * View of the head of Eigen vector types.
+   * @param n Number of elements to return from top of vector.
+   */
+  inline const auto head(Eigen::Index n) const {
+    return vari_value<const PlainObject>(val_.head(n), adj_.head(n));
+  }
+
+  /**
+   * View of the tail of the Eigen vector types.
+   * @param n Number of elements to return from bottom of vector.
+   */
+  inline const auto tail(Eigen::Index n) const {
+    return vari_value<const PlainObject>(val_.tail(n), adj_.tail(n));
+  }
+
+  /**
+   * View block of N elements starting at position `i`
+   * @param i Starting position of block.
+   * @param n Number of elements in block
+   */
+  inline const auto segment(Eigen::Index i, Eigen::Index n) const {
+    return vari_value<const PlainObject>(val_.segment(i, n),
+                                         adj_.segment(i, n));
+  }
+
+  /**
+   * View row of eigen matrices.
+   * @param i Row index to slice.
+   */
+  inline const auto row(Eigen::Index i) const {
+    const auto& val_row = val_.row(i);
+    const auto& adj_row = adj_.row(i);
+    return vari_value<const PlainObject, Eigen::InnerStride<>>(
+        val_row, adj_row, val_row.innerStride(), adj_row.innerStride());
+  }
+
+  /**
+   * View column of eigen matrices
+   * @param i Column index to slice
+   */
+  inline const auto col(Eigen::Index i) const {
+    const auto& val_col = val_.col(i);
+    const auto& adj_col = adj_.col(i);
+    return vari_value<const PlainObject, Eigen::OuterStride<>>(
+        val_col, adj_col, val_col.outerStride(), adj_col.outerStride());
+  }
+
+  /**
+   * Get coefficient of eigen matrices
+   * @param i Row index
+   * @param j Column index
+   */
+  inline const auto coeff(Eigen::Index i, Eigen::Index j) const {
+    return vari_value<double>(val_(i, j), adj_(i, j));
```

Comment:
Making a copy of the value and adjoint is not the correct thing to do here, as setting adjoint of the copy will not propagate it back to the matrix.

Reply:
Yes that's intented since you can't assign to subsets of a `var_value<>`. That's also why these are all returning const types

Reply:
Yes so I agree with this, what I'm trying to figure out right now is that for the var to decide if it holds a view or an actual `vari_value`  I'm doing 

```
  using vari_type = std::conditional_t<is_plain_type<value_type>::value,
                                       vari_value<value_type>, vari_view<T>>;
```

So this is fine and good for `Block` types and other views, but for views into coefficients `value_type`/`T` will be a `double` which is the plain type so this will become a `vari_value<double>`. We don't want to explicitly have a second template parameter to set the `vari_type` to, can you think of another way to handle this?

Reply:
I would make a separate class `var_view` (not `vari`). This also avoids two pointer dereferences (one to get vari and second to get value/adjoint) )when operationg on such a scalar. Also since this class is only used for scalars we don't need to template it.

We could also make this object be implicitly convertible to `var_value<double>`, registering the callback to propagate adjoint on the stack. Although for performance reason we could also not (or at least make it explicit) and explicitly use `var_view` wherever this is used.

Reply:
Interesting! Okay I'll work on this today

Reply:
So I had a bit of a hard time getting what you suggested to work. In `var<mat>.coeff` I added a callback to propogate the adjoint of the `var_value<double>` back to the original `var<mat>`. I think this makes since assuming that the two cases this will be used 99% of the time are

1. assigning a cell of a static matrix to a var

```stan
real a = -B[i];
```

2. iterating over a static matrix to assign to the cells in a dynamic matrix

```stan
matrix[N, M] A; // assume static
matrix[N, M] B; // assume dynamic
for (i in 1:N) {
  B[i] = -A[i]; // very inefficient but a thing a user might want to do
}
```

so for both of these we will need to make a var anyway so having a `var_view` would not help. Are there normal cases I'm missing that a `var_view` would be more efficient? If you have any time to take a crack at this I'd be interested in seeing what you have in mind!